packages/wrangler/src/experimental-assets.ts

```diff
@@ -1,8 +1,299 @@
+					1
+				);
+			}
+			// just logging file uploads at the moment...
+			// unsure how to log deletion vs unchanged file ignored/if we want to log this
+			assetLogCount = logAssetUpload(`+ ${manifestEntry[0]}`, assetLogCount);
+			return manifestEntry;
+		});
+	});
+
+	const queue = new PQueue({ concurrency: BULK_UPLOAD_CONCURRENCY });
+	let attempts = 0;
+	const start = Date.now();
+	let completionJwt = "";
+
+	for (const [bucketIndex, bucketSkeleton] of bucketSkeletons.entries()) {
+		attempts = 0;
+		let gatewayErrors = 0;
+		const doUpload = async (): Promise<UploadResponse> => {
+			// Populate the payload only when actually uploading (this is limited to 3 concurrent uploads at 50 MiB per bucket meaning we'd only load in a max of ~150 MiB)
+			// This is so we don't run out of memory trying to upload the files.
+			const payload: UploadPayloadFile[] = await Promise.all(
+				bucketSkeleton.map(async (manifestEntry) => {
+					const absFilePath = path.join(assetDirectory, manifestEntry[0]);
+					return {
+						base64: true,
+						key: manifestEntry[1].hash,
+						metadata: {
+							contentType: getType(absFilePath) || "application/octet-stream",
+						},
+						value: (await readFile(absFilePath)).toString("base64"),
+					};
+				})
+			);
+
+			try {
+				const res = await fetchResult<UploadResponse>(
+					`/accounts/${accountId}/workers/assets/upload`,
+					{
+						method: "POST",
+						headers: {
+							"Content-Type": "application/x-ndjson",
+							Authorization: `Bearer ${initializeAssetsResponse.jwt}`,
+						},
+						body: payload.map((x) => JSON.stringify(x)).join("\n"),
+					}
+				);
+				logger.info(
+					`Uploaded bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length}`
+				);
+				return res;
+			} catch (e) {
+				if (attempts < MAX_UPLOAD_ATTEMPTS) {
+					logger.info(
+						chalk.dim(
+							`Bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length} upload failed. Retrying...\n`,
+							e
+						)
+					);
+					// Exponential backoff, 1 second first time, then 2 second, then 4 second etc.
+					await new Promise((resolvePromise) =>
+						setTimeout(resolvePromise, Math.pow(2, attempts) * 1000)
+					);
+					if (e instanceof APIError && e.isGatewayError()) {
+						// Gateway problem, wait for some additional time and set concurrency to 1
+						queue.concurrency = 1;
+						await new Promise((resolvePromise) =>
+							setTimeout(resolvePromise, Math.pow(2, gatewayErrors) * 5000)
+						);
+						gatewayErrors++;
+						// only count as a failed attempt after a few initial gateway errors
+						if (gatewayErrors >= MAX_UPLOAD_GATEWAY_ERRORS) {
+							attempts++;
+						}
+					} else {
+						attempts++;
+					}
+					return doUpload();
+				} else if (isJwtExpired(initializeAssetsResponse.jwt)) {
+					throw new FatalError(
+						`Asset upload took too long on bucket ${bucketIndex + 1}/${initializeAssetsResponse.buckets.length}. Please try again.
```

Comment:
I noticed that we're using `assets` / `files` interchangeably in our copies. Perhaps let's settle on one for consistency purposes?

Reply:
[followup ticket](https://jira.cfdata.org/browse/RM-19435)